gobblin-core/src/main/java/gobblin/source/extractor/filebased/FileBasedSource.java

```diff
@@ -92,14 +92,17 @@ protected void initLogger(SourceState state) {
     }
 
     TableType tableType = TableType.valueOf(state.getProp(ConfigurationKeys.EXTRACT_TABLE_TYPE_KEY).toUpperCase());
-    List<WorkUnitState> previousWorkunits = Lists.newArrayList(state.getPreviousWorkUnitStates());
+    SourceState previousSourceState = state.getPreviousSourceState();
+
     List<String> prevFsSnapshot = Lists.newArrayList();
 
     // Get list of files seen in the previous run
-    if (!previousWorkunits.isEmpty()
-        && previousWorkunits.get(0).getWorkunit().contains(ConfigurationKeys.SOURCE_FILEBASED_FS_SNAPSHOT)) {
-      prevFsSnapshot =
-          previousWorkunits.get(0).getWorkunit().getPropAsList(ConfigurationKeys.SOURCE_FILEBASED_FS_SNAPSHOT);
+    if (previousSourceState != null) {
+      state.setProp(ConfigurationKeys.PREVIOUS_STATE_ID_KEY,
+          previousSourceState.getProp(ConfigurationKeys.DATASET_STATE_ID_KEY, ConfigurationKeys.DEFAULT_DATASET_STATE_ID));
+      if (previousSourceState.contains(ConfigurationKeys.SOURCE_FILEBASED_FS_SNAPSHOT)) {
+        prevFsSnapshot = previousSourceState.getPropAsList(ConfigurationKeys.SOURCE_FILEBASED_FS_SNAPSHOT);
```

Comment:
I think this is still needed, so I would love a suggestion on how to accomplish this.  The reason I want this is that it makes it *very* easy to diagnose issues where jobs are running off of the same previous state.  This scenario is particularly insidious as it results in duplicated data.